### PR TITLE
Adding scipy.signal.convolve and correlate functions

### PR DESCRIPTION
**Pull request by coderforlife**
This adds multidimensional `cupyx.scipy.signal.convolve()` and `correlate()` functions. These simply map to the `ndimage` functions (which were recently prepared for this with added support for the `full` and `valid` modes natively without doing any extra padding).  

For the moment these functions aren't very useful since by default `method='auto'` and `choose_conv_method()` always returns `'fft'` which isn't in this PR. However, `method='direct'` is fully supported (including complex types even though #3705 isn't resolved since it doesn't use `_run_1d_correlates()`).

I also have the code ready to submit for **`fftconvolve()`** and  **`choose_conv_method()`** which I will create a PR for after this one is accepted (they are a bit intertwined, easier to separate this one out than that one). After that, I have `convolve2`, `correlate2`, `oaconvolve` all ready to go as well.

This will need some minor rebasing after #3743 is merged. Additionally, I will update this PR in the next day or two with tests (have done informal tests and they all worked).

### CODE REVIEW

**Comment by coderforlife** [comment]
I decided to add `convolve2d` and `correlate2d` as well since the add very little code above what `convolve` and `correlate` add.

**Comment by coderforlife** [comment]
I have written most of the tests and in doing so found a bug in scipy: https://github.com/scipy/scipy/issues/12686. I will avoid testing against that situation for now.

**Comment by coderforlife** [comment]
This commit would end up closing #1639 (which asks for the 4 functions this PR adds).

**Comment by coderforlife** [comment]
Not that complex types apparently aren't working as well as I thought. Simple cases with complex do work, but any random matrix doesn't. Working on that issue now.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
Would you add the reference entries in `cupy/docs/source/reference/signal.rst`?

File: cupyx/scipy/signal/signaltools.py [hunk 1]
```diff
@@ -47,6 +156,109 @@ def _fftconv_faster(x, h, mode):
...
+        raise ValueError('{} inputs must both be 2-D arrays'.format(
+            'convolve2d' if convolution else 'correlate2d'))
+    _boundaries = {
+        'fill': 'constant', 'pad': 'constant',
+        'wrap': 'wrap', 'circular': 'wrap',
+        'symm': 'reflect', 'symmetric': 'reflect',
+    }
+    boundary = _boundaries.get(boundary)
+    if boundary is None:
+        raise ValueError("Acceptable boundary flags are 'fill' (or 'pad'), "
```
Comment (takagi) [inline cupyx/scipy/signal/signaltools.py:252]: Why not `fill`, `wrap` and `symm` only here?

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 2]
```diff
@@ -13,6 +13,94 @@
...
+    def test_correlate_float16(self, xp, scp, dtype=cupy.float16):
+        return self._filter('correlate', dtype, xp, scp)
+
+
+@testing.parameterize(*testing.product({
+    'in1': [(5, 10), (10, 7)],
+    'in2': [(3, 2), (3, 3), (2, 2), (10, 10), (11, 11)],
+    'mode': ['full', 'same', 'valid'],
+    'boundary': ['fill', 'wrap', 'symm'],
+}))
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:63]: Would you test `fillvalue` as well?

File: cupyx/scipy/signal/signaltools.py [hunk 3]
```diff
@@ -47,6 +156,109 @@ def _fftconv_faster(x, h, mode):
...
+    Cross correlate ``in1`` and ``in2`` with output size determined by
+    ``mode``, and boundary conditions determined by ``boundary`` and
+    ``fillvalue``.
+
+    Args:
+        in1 (cupy.ndarray): First input.
+        in2 (cupy.ndarray): Second input. Should have the same number of
+            dimensions as ``in1``.
+        mode (str): Indicates the size of the output:
+            ``'full'``: output is the full discrete linear convolution
```
Comment (takagi) [inline cupyx/scipy/signal/signaltools.py:210]:
```suggestion
            ``'full'``: output is the full discrete linear cross-correlation
```

File: cupyx/scipy/signal/_signaltools_core.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            offsets = tuple((x-shift)//2 for x in in2.shape)
+        else:
+            offsets = tuple((2*x2-x1-(not convolution)+shift)//2
+                            for x1, x2 in zip(in1.shape, in2.shape))
+
+    # Check the output
+    if not isinstance(output, cupy.ndarray):
+        output = cupy.empty(out_shape, output)
+    elif output.shape != out_shape:
+        print(output.shape, out_shape)
```
Comment (takagi) [inline cupyx/scipy/signal/_signaltools_core.py:63]:
```suggestion
```

File: cupyx/scipy/signal/signaltools.py (hunk 1 above)
Comment (coderforlife) [reply cupyx/scipy/signal/signaltools.py:252]: This mimics the scipy error message which includes the un-documented acceptable boundary flags.

File: cupyx/scipy/signal/_signaltools_core.py (hunk 4 above)
Comment (coderforlife) [reply cupyx/scipy/signal/_signaltools_core.py:63]: Oops! Thanks for catching that.

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
pfnCI, test this please.

**Pull request merged**